### PR TITLE
feat(Plugin): new lifecycle methods

### PR DESCRIPTION
**Pull request by DiamondMiner88**
time for an overhaul

- Deprecate `load(context)` in favor of new `onLoad()`
- Deprecate `unload(context)` in favor of new `onUnload()`
- Deprecate `start(context)` in favor of new `onStart()`
- Deprecate `stop(context)` in favor of new `onStop()`
- New `onInstall()`
- New `onFirstInstall()`
- New `onBeforeUpdate(nextVersion)`
- New `onAfterUpdate(prevVersion)`
- New `onUninstall()`

### CODE REVIEW

File: Aliucord/src/main/java/com/aliucord/entities/Plugin.java [hunk 1]
```diff
@@ -168,28 +169,104 @@ public AliucordPlugin getAnnotation() {
...
+     */
+    @SuppressWarnings("RedundantThrows")
+    public void onInstall() throws Throwable {}
+
+    /**
+     * Called when your plugin is installed for the first time
+     * Called after {@link Plugin#onInstall()}
+     */
+    @SuppressWarnings("RedundantThrows")
+    public void onFirstInstall() throws Throwable {}
```
Comment (Vendicated) [inline Aliucord/src/main/java/com/aliucord/entities/Plugin.java:249]: What would be the use case for this?

File: Aliucord/src/main/java/com/aliucord/settings/Plugins.java [hunk 2]
```diff
@@ -239,20 +239,26 @@ public void onToggleClick(ViewHolder holder, boolean state, int position) {
...
-                PluginManager.stopPlugin(p.getName());
-                PluginManager.plugins.remove(p.getName());
-                PluginManager.logger.info(ctx, "Successfully deleted " + p.getName());
+                try {
+                    p.onUninstall();
+                } catch (Throwable err) { PluginManager.logger.error("Exception while uninstalling plugin: " + name, err); }
+
+                PluginManager.stopPlugin(name);
+                PluginManager.plugins.remove(name);
+                SettingsUtils.delete(PluginManager.getPluginPrefKey(name));
```
Comment (Vendicated) [inline Aliucord/src/main/java/com/aliucord/settings/Plugins.java:260]: why did you add this?

File: Aliucord/src/main/java/com/aliucord/updater/PluginUpdater.java [hunk 3]
```diff
@@ -170,6 +174,10 @@ public static boolean update(String plugin) throws Throwable {
             );
         }
 
+        try {
+            p.onAfterUpdate(p.getManifest().version);
```
Comment (Vendicated) [inline Aliucord/src/main/java/com/aliucord/updater/PluginUpdater.java:178]: i think this should be called after the remount so it uses the method of the newly updated plugin instead of the old one

**Comment by Vendicated** [comment]
solve conflicts please

File: Aliucord/src/main/java/com/aliucord/updater/PluginUpdater.java (hunk 3 above)
Comment (DiamondMiner88) [reply Aliucord/src/main/java/com/aliucord/updater/PluginUpdater.java:178]: Issue is that it might not always be triggered then (for example when the plugin is disabled when updated)

File: Aliucord/src/main/java/com/aliucord/updater/PluginUpdater.java [hunk 4]
```diff
@@ -174,17 +174,24 @@ public static boolean update(String plugin) throws Throwable {
             );
         }
 
-        try {
-            p.onAfterUpdate(p.getManifest().version);
-        } catch (Throwable e) { logger.error("Exception while updating plugin: " + p.getName(), e); }
+        Utils.mainThread.post(() -> {
+            if (PluginManager.isPluginEnabled(plugin))
+                PluginManager.disablePlugin(plugin);
```
Comment (Vendicated) [inline Aliucord/src/main/java/com/aliucord/updater/PluginUpdater.java:179]: ??

Comment (Vendicated) [reply Aliucord/src/main/java/com/aliucord/updater/PluginUpdater.java:181]: why do you disable the plugin ???

File: Aliucord/src/main/java/com/aliucord/settings/Plugins.java (hunk 2 above)
Comment (DiamondMiner88) [reply Aliucord/src/main/java/com/aliucord/settings/Plugins.java:260]: When deleting disabled plugins and reinstalling later it's annoying to have to go enable it again

Comment (ItzOnlyAnimal) [reply Aliucord/src/main/java/com/aliucord/settings/Plugins.java:260]: wouldn't this remove it's setting also though?

File: Aliucord/src/main/java/com/aliucord/updater/PluginUpdater.java (hunk 4 above)
Comment (DiamondMiner88) [reply Aliucord/src/main/java/com/aliucord/updater/PluginUpdater.java]: Fixed now, I think

File: Aliucord/src/main/java/com/aliucord/PluginManager.java [hunk 5]
```diff
@@ -152,9 +153,15 @@ public static void togglePlugin(String name) {
...
         logger.info("Starting plugin: " + name);
-         try {
-            Objects.requireNonNull(plugins.get(name)).start(Utils.getAppContext());
-         } catch (Throwable e) { logger.error("Exception while starting plugin: " + name, e); }
+        try {
+            Plugin p = Objects.requireNonNull(plugins.get(name));
+            p.onStart();
+            if (!SettingsUtils.exists(getPluginPrefKey(name))) {
+                SettingsUtils.setBool(getPluginPrefKey(name), true);
+                p.onInstall();
```
Comment (Vendicated) [inline Aliucord/src/main/java/com/aliucord/PluginManager.java:161]: only called on first install, should be called on all installs

File: Aliucord/src/main/java/com/aliucord/entities/Plugin.java [hunk 6]
```diff
@@ -182,28 +184,104 @@ public AliucordPlugin getAnnotation() {
...
-    public abstract void stop(Context context) throws Throwable;
+    @Deprecated
+    @SuppressWarnings("RedundantThrows")
+    public void stop(Context context) throws Throwable {}
+
+    /**
+     * Called when your plugin is stopped
+     */
+    @SuppressWarnings("deprecation")
+    public void onStop() throws Throwable {
```
Comment (Vendicated) [inline Aliucord/src/main/java/com/aliucord/entities/Plugin.java:248]: should be abstract to enforce implementing this method

**Review by Vendicated — CHANGES_REQUESTED** [review]
